File: internal/dao/membership_plan.go
package dao

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

// MembershipPlan represents a plan that a gym offers to its members.
// BillingType is "recurring" or "one_time".
// Interval is "monthly", "yearly", or "weekly" (only applies when BillingType == "recurring").
// Price is stored in cents (e.g. 10000 = $100.00).
// ClassLimit nil means unlimited classes.
type MembershipPlan struct {
	ID          bson.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
	GymID       bson.ObjectID `json:"gym_id" bson:"gym_id"`
	Name        string        `json:"name" bson:"name" validate:"required"`
	Description string        `json:"description" bson:"description"`
	BillingType string        `json:"billing_type" bson:"billing_type" validate:"required,oneof=recurring one_time"`
	Interval    string        `json:"interval" bson:"interval"`
	Price       int64         `json:"price" bson:"price" validate:"required,min=0"`
	Currency    string        `json:"currency" bson:"currency"`
	ClassLimit  *int          `json:"class_limit" bson:"class_limit"`
	IsActive    bool          `json:"is_active" bson:"is_active"`
	CreatedAt   time.Time     `json:"created_at" bson:"created_at"`
	UpdatedAt   time.Time     `json:"updated_at" bson:"updated_at"`
}

File: grapple-frontend/src/api-requests/membershipPlan.ts
import { membershipPlansApi } from "@/hook/base-apis";

export interface MembershipPlan {
  id?: string;
  gym_id: string;
  name: string;
  description: string;
  billing_type: 'recurring' | 'one_time';
  interval?: 'monthly' | 'yearly' | 'weekly';
  price: number;
  currency: string;
  class_limit: number | null;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
}

const authHeaders = (token: string) => ({
  headers: { Authorization: `Bearer ${token}` },
});

export const getMembershipPlans = async (gymId: string, token: string): Promise<MembershipPlan[]> => {
  const { data } = await membershipPlansApi.get<MembershipPlan[]>('', {
    params: { gym_id: gymId },
    ...authHeaders(token),
  });
  return data;
};

export const createMembershipPlan = async (plan: Omit<MembershipPlan, 'id' | 'is_active' | 'created_at' | 'updated_at'>, token: string): Promise<MembershipPlan> => {
  const { data } = await membershipPlansApi.post<MembershipPlan>('', plan, authHeaders(token));
  return data;
};

export const updateMembershipPlan = async (gymId: string, planId: string, plan: Partial<MembershipPlan>, token: string): Promise<MembershipPlan> => {
  const { data } = await membershipPlansApi.put<MembershipPlan>(`/${gymId}/${planId}`, plan, authHeaders(token));
  return data;
};

export const deleteMembershipPlan = async (gymId: string, planId: string, token: string): Promise<void> => {
  await membershipPlansApi.delete(`/${gymId}/${planId}`, authHeaders(token));
};
